File: MiszIQAndroid/app/src/main/java/com/example/misziq/ui/components/DifficultySelector.kt
package com.example.misziq.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.misziq.data.model.*
import com.example.misziq.data.repository.MiszIQRepository
import com.example.misziq.ui.theme.RoyalBlue
import kotlinx.coroutines.launch

@Composable
fun DifficultySelector(
    profileId: String,
    gameType: GameType,
    repository: MiszIQRepository,
    selectedLevel: Int,
    onLevelSelected: (Int) -> Unit,
    onStart: () -> Unit
) {
    var maxUnlockedLevel by remember { mutableIntStateOf(1) }
    val scope = rememberCoroutineScope()

    LaunchedEffect(profileId, gameType) {
        maxUnlockedLevel = repository.getMaxUnlockedLevel(profileId, gameType.name)
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(
            text = "Select Difficulty",
            style = MaterialTheme.typography.titleLarge,
            fontWeight = FontWeight.Bold
        )

        Spacer(modifier = Modifier.height(24.dp))

        Column(
            verticalArrangement = Arrangement.spacedBy(12.dp),
            modifier = Modifier.fillMaxWidth()
        ) {
            listOf(
                Triple(1, "Easy", "Great for beginners"),
                Triple(2, "Medium", "A balanced challenge"),
                Triple(3, "Hard", "For experienced players")
            ).forEach { (level, name, description) ->
                val isUnlocked = level <= maxUnlockedLevel
                val isSelected = selectedLevel == level

                DifficultyRow(
                    level = level,
                    name = name,
                    description = description,
                    isSelected = isSelected,
                    isUnlocked = isUnlocked,
                    onClick = {
                        if (isUnlocked) {
                            onLevelSelected(level)
                        }
                    }
                )
            }
        }

        if (maxUnlockedLevel < 3) {
            Spacer(modifier = Modifier.height(16.dp))
            Text(
                text = "Achieve 100% accuracy to unlock higher levels",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }

        Spacer(modifier = Modifier.height(24.dp))

        Button(
            onClick = onStart,
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 40.dp),
            colors = ButtonDefaults.buttonColors(containerColor = RoyalBlue)
        ) {
            Text("Start Game", style = MaterialTheme.typography.titleMedium)
        }
    }
}

@Composable
fun DifficultyRow(
    level: Int,
    name: String,
    description: String,
    isSelected: Boolean,
    isUnlocked: Boolean,
    onClick: () -> Unit
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(12.dp))
            .background(
                if (isSelected && isUnlocked) RoyalBlue.copy(alpha = 0.1f)
                else MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.5f)
            )
            .border(
                width = if (isSelected && isUnlocked) 2.dp else 0.dp,
                color = if (isSelected && isUnlocked) RoyalBlue else Color.Transparent,
                shape = RoundedCornerShape(12.dp)
            )
            .clickable(enabled = isUnlocked) { onClick() }
            .padding(16.dp),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Column {
            Text(
                text = name,
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
                color = if (isUnlocked) MaterialTheme.colorScheme.onSurface
                        else MaterialTheme.colorScheme.onSurfaceVariant
            )
            Text(
                text = description,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }

        if (isUnlocked) {
            if (isSelected) {
                Icon(
                    imageVector = Icons.Default.Check,
                    contentDescription = "Selected",
                    tint = RoyalBlue,
                    modifier = Modifier.size(28.dp)
                )
            }
        } else {
            Icon(
                imageVector = Icons.Default.Lock,
                contentDescription = "Locked",
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.size(28.dp)
            )
        }
    }
}

@Composable
fun GameCompletionOverlay(
    newBadges: List<BadgeType>,
    unlockedLevel: Int?,
    gameType: GameType,
    onDismiss: () -> Unit
) {
    if (newBadges.isNotEmpty() || unlockedLevel != null) {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(Color.Black.copy(alpha = 0.5f))
                .clickable(enabled = false) { },
            contentAlignment = Alignment.Center
        ) {
            Card(
                modifier = Modifier
                    .fillMaxWidth(0.85f)
                    .padding(16.dp),
                shape = RoundedCornerShape(20.dp)
            ) {
                Column(
                    modifier = Modifier.padding(24.dp),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    if (newBadges.isNotEmpty()) {
                        Text(
                            text = "Badge Unlocked!",
                            style = MaterialTheme.typography.titleLarge,
                            fontWeight = FontWeight.Bold,
                            color = RoyalBlue
                        )

                        Spacer(modifier = Modifier.height(16.dp))

                        newBadges.forEach { badge ->
                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .background(
                                        RoyalBlue.copy(alpha = 0.1f),
                                        RoundedCornerShape(12.dp)
                                    )
                                    .padding(12.dp),
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Text(
                                    text = badge.emoji,
                                    fontSize = 36.sp
                                )
                                Spacer(modifier = Modifier.width(12.dp))
                                Column {
                                    Text(
                                        text = badge.displayName,
                                        style = MaterialTheme.typography.titleMedium,
                                        fontWeight = FontWeight.Bold
                                    )
                                    Text(
                                        text = badge.description,
                                        style = MaterialTheme.typography.bodySmall,
                                        color = MaterialTheme.colorScheme.onSurfaceVariant
                                    )
                                }
                            }
                            Spacer(modifier = Modifier.height(8.dp))
                        }
                    }

                    if (unlockedLevel != null) {
                        if (newBadges.isNotEmpty()) {
                            Spacer(modifier = Modifier.height(16.dp))
                        }

                        Column(
                            modifier = Modifier
                                .fillMaxWidth()
                                .background(
                                    Color(0xFF40E0D0).copy(alpha = 0.1f),
                                    RoundedCornerShape(12.dp)
                                )
                                .padding(16.dp),
                            horizontalAlignment = Alignment.CenterHorizontally
                        ) {
                            Text(text = "🔓", fontSize = 40.sp)
                            Spacer(modifier = Modifier.height(8.dp))
                            Text(
                                text = "Level $unlockedLevel Unlocked!",
                                style = MaterialTheme.typography.titleMedium,
                                fontWeight = FontWeight.Bold
                            )
                            Text(
                                text = "You can now play ${gameType.displayName} on a harder difficulty!",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                    }

                    Spacer(modifier = Modifier.height(24.dp))

                    Button(
                        onClick = onDismiss,
                        colors = ButtonDefaults.buttonColors(containerColor = RoyalBlue)
                    ) {
                        Text("Continue")
                    }
                }
            }
        }
    }
}

File: MiszIQ/DifficultySelector.swift
import SwiftUI
import SwiftData

struct DifficultySelector: View {
    let profile: UserProfile
    let gameType: GameType
    @Binding var selectedLevel: Int
    let onStart: () -> Void

    private var maxUnlockedLevel: Int {
        BadgeManager.getMaxUnlockedLevel(profile: profile, gameType: gameType)
    }

    var body: some View {
        VStack(spacing: 24) {
            Text("Select Difficulty")
                .font(.title2.bold())

            VStack(spacing: 12) {
                ForEach(1...3, id: \.self) { level in
                    let isUnlocked = level <= maxUnlockedLevel
                    DifficultyRow(
                        level: level,
                        isSelected: selectedLevel == level,
                        isUnlocked: isUnlocked
                    ) {
                        if isUnlocked {
                            selectedLevel = level
                        }
                    }
                }
            }
            .padding(.horizontal)

            if maxUnlockedLevel < 3 {
                Text("Achieve 100% accuracy to unlock higher levels")
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .multilineTextAlignment(.center)
            }

            Button {
                onStart()
            } label: {
                Text("Start Game")
                    .font(.headline)
                    .frame(maxWidth: .infinity)
                    .padding()
                    .background(Color.royalBlue)
                    .foregroundStyle(.white)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
            }
            .padding(.horizontal, 40)
        }
    }
}

struct DifficultyRow: View {
    let level: Int
    let isSelected: Bool
    let isUnlocked: Bool
    let onTap: () -> Void

    private var levelName: String {
        switch level {
        case 1: return "Easy"
        case 2: return "Medium"
        case 3: return "Hard"
        default: return "Level \(level)"
        }
    }

    private var levelDescription: String {
        switch level {
        case 1: return "Great for beginners"
        case 2: return "A balanced challenge"
        case 3: return "For experienced players"
        default: return ""
        }
    }

    var body: some View {
        Button(action: onTap) {
            HStack {
                VStack(alignment: .leading, spacing: 4) {
                    Text(levelName)
                        .font(.headline)
                        .foregroundStyle(isUnlocked ? .primary : .secondary)

                    Text(levelDescription)
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }

                Spacer()

                if isUnlocked {
                    if isSelected {
                        Image(systemName: "checkmark.circle.fill")
                            .foregroundStyle(Color.royalBlue)
                            .font(.title2)
                    } else {
                        Image(systemName: "circle")
                            .foregroundStyle(.secondary)
                            .font(.title2)
                    }
                } else {
                    Image(systemName: "lock.fill")
                        .foregroundStyle(.secondary)
                        .font(.title2)
                }
            }
            .padding()
            .background(
                RoundedRectangle(cornerRadius: 12)
                    .fill(isSelected && isUnlocked ? Color.royalBlue.opacity(0.1) : Color.gray.opacity(0.1))
            )
            .overlay(
                RoundedRectangle(cornerRadius: 12)
                    .stroke(isSelected && isUnlocked ? Color.royalBlue : Color.clear, lineWidth: 2)
            )
        }
        .buttonStyle(.plain)
        .disabled(!isUnlocked)
    }
}

// Helper to handle badge checking and difficulty unlocks after game completion
struct GameCompletionHandler {
    static func handleGameCompletion(
        profile: UserProfile,
        gameType: GameType,
        score: Int,
        maxPossibleScore: Int,
        level: Int,
        durationSeconds: Int,
        mockService: MockDataService,
        modelContext: ModelContext
    ) -> (newBadges: [BadgeType], unlockedLevel: Int?) {
        // Create and save the session
        let session = GameSession(
            gameType: gameType,
            score: score,
            maxPossibleScore: maxPossibleScore,
            level: level,
            durationSeconds: durationSeconds
        )
        session.profile = profile
        modelContext.insert(session)

        // Check for new badges
        let newBadges = BadgeManager.checkForNewBadges(
            profile: profile,
            session: session,
            mockService: mockService,
            modelContext: modelContext
        )

        // Check for difficulty unlock
        let unlockedLevel = BadgeManager.checkDifficultyUnlock(
            profile: profile,
            session: session,
            modelContext: modelContext
        )

        return (newBadges, unlockedLevel)
    }
}

// A view to show unlocked badges and difficulty levels
struct GameCompletionOverlay: View {
    let newBadges: [BadgeType]
    let unlockedLevel: Int?
    let gameType: GameType
    let onDismiss: () -> Void

    var body: some View {
        if !newBadges.isEmpty || unlockedLevel != nil {
            VStack(spacing: 20) {
                if !newBadges.isEmpty {
                    VStack(spacing: 12) {
                        Text("Badge Unlocked!")
                            .font(.title2.bold())
                            .foregroundStyle(Color.royalBlue)

                        ForEach(newBadges, id: \.self) { badge in
                            HStack(spacing: 12) {
                                Text(badge.emoji)
                                    .font(.system(size: 40))

                                VStack(alignment: .leading) {
                                    Text(badge.displayName)
                                        .font(.headline)
                                    Text(badge.description)
                                        .font(.caption)
                                        .foregroundStyle(.secondary)
                                }
                            }
                            .padding()
                            .background(Color.royalBlue.opacity(0.1))
                            .clipShape(RoundedRectangle(cornerRadius: 12))
                        }
                    }
                }

                if let level = unlockedLevel {
                    VStack(spacing: 8) {
                        Image(systemName: "lock.open.fill")
                            .font(.system(size: 40))
                            .foregroundStyle(Color.turquoise)

                        Text("Level \(level) Unlocked!")
                            .font(.headline)

                        Text("You can now play \(gameType.rawValue) on a harder difficulty!")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                            .multilineTextAlignment(.center)
                    }
                    .padding()
                    .background(Color.turquoise.opacity(0.1))
                    .clipShape(RoundedRectangle(cornerRadius: 12))
                }

                Button("Continue", action: onDismiss)
                    .font(.headline)
                    .padding(.horizontal, 40)
                    .padding(.vertical, 12)
                    .background(Color.royalBlue)
                    .foregroundStyle(.white)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
            }
            .padding(24)
            .background(
                RoundedRectangle(cornerRadius: 20)
                    .fill(Color(UIColor.systemBackground))
                    .shadow(radius: 20)
            )
            .padding(40)
        }
    }
}
